File: solutions/2025-12-02/solution.py
# Validate IP Address
# Difficulty: Medium
# URL: https://leetcode.com/problems/validate-ip-address/

class Solution:
    def validIPAddress(self, IP: str) -> str:
        parts = IP.split('.')
        if len(parts) != 4:
            return "Neither"
        for part in parts:
            if len(part) > 1 and part[0] == "0":
                return "Neither"
            if not part.isdigit() or not 0 <= int(part) <= 255:
                return "Neither"
            if len(part) > 3:
                return "Neither"
        return "IPv4"
    def validIPv6(self, IP: str) -> str:
        parts = IP.split(':')
        if len(parts) != 8:
            return "Neither"
        for part in parts:
            if len(part) > 4:
                return "Neither"
            for char in part:
                if not char.isdigit() and not (char.lower() >= 'a' and char.lower() <= 'f'):
                    return "Neither"
        return "IPv6"

File: solutions/2025-12-02/solution.java
// Validate IP Address
// Difficulty: Medium
// URL: https://leetcode.com/problems/validate-ip-address/

class Solution {
    public String validIPAddress(String IP) {
        String[] parts;
        if (IP.contains(".")) {
            parts = IP.split(".");
            if (parts.length != 4) {
                return "Neither";
            }
            for (String part : parts) {
                if (part.length() > 1 && part.charAt(0) == '0') {
                    return "Neither";
                }
                if (!part.matches("[0-9]+") || Integer.parseInt(part) > 255 || Integer.parseInt(part) < 0) {
                    return "Neither";
                }
                if (part.length() > 3) {
                    return "Neither";
                }
            }
            return "IPv4";
        } else if (IP.contains(":")) {
            parts = IP.split(":");
            if (parts.length != 8) {
                return "Neither";
            }
            for (String part : parts) {
                if (part.length() > 4) {
                    return "Neither";
                }
                for (char c : part.toCharArray()) {
                    if (!Character.isDigit(c) && !(c >= 'a' && c <= 'f') && !(c >= 'A' && c <= 'F')) {
                        return "Neither";
                    }
                }
            }
            return "IPv6";
        } else {
            return "Neither";
        }
    }
}